File: day02/part1.c
#include <assert.h>
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>

long compute(long* prog) {
    size_t pc = 0;
    while (true) {
        long opc = prog[pc];
        if (opc == 99) {
            return prog[0];
        } else if (opc == 1) {
            prog[prog[pc + 3]] = prog[prog[pc + 2]] + prog[prog[pc + 1]];
        } else if (opc == 2) {
            prog[prog[pc + 3]] = prog[prog[pc + 2]] * prog[prog[pc + 1]];
        } else {
            assert(!"unreachable");
        }
        pc += 4;
    }
    assert(!"unreachable");
}

int main(int argc, char* argv[]) {
    if (argc != 2) {
        fprintf(stderr, "usage: %s FILENAME\n", argv[0]);
        return 1;
    }

    FILE* f = fopen(argv[1], "r");
    if (f == NULL) {
        fprintf(stderr, "invalid filename: %s\n", argv[1]);
        return 1;
    }

    size_t size = 4;
    size_t idx = 0;
    long* arr = malloc(sizeof *arr * size);

    if (fscanf(f, "%ld", arr + idx) != 1) {
        fprintf(stderr, "expected numbers????\n");
        fclose(f);
        return 1;
    }
    idx += 1;

    while (fscanf(f, ",%ld", arr + idx) == 1) {
        idx += 1;
        if (idx >= size) {
            long* new_arr = malloc(sizeof *new_arr * size * 2);
            memcpy(new_arr, arr, sizeof *arr * size);
            free(arr);
            arr = new_arr;
            size *= 2;
        }
    }

    arr[1] = 12;
    arr[2] = 2;
    printf("%ld\n", compute(arr));

    free(arr);
    fclose(f);
    return 0;
}

File: day02/part1.js
'use strict';

const fs = require('fs');

if (process.argv.length !== 3) {
    console.error(`usage: node ${process.argv[1]} FILENAME`);
    process.exit(1)
}


function compute(prog) {
    let pc = 0;
    for (;;) {
        const opc = prog[pc];
        switch (opc) {
            case 99:
                return prog[0];
            case 1:
                prog[prog[pc + 3]] = prog[prog[pc + 2]] + prog[prog[pc + 1]];
                break;
            case 2:
                prog[prog[pc + 3]] = prog[prog[pc + 2]] * prog[prog[pc + 1]];
                break;
            default:
                throw new Error('OHNOES!');
        }
        pc += 4;
    }
}

let prog =
    fs.readFileSync(process.argv[2], {encoding: 'UTF-8'})
        .trim()
        .split(/,/)
        .map((s) => parseInt(s, 10));

prog[1] = 12;
prog[2] = 2;

console.log(compute(prog));
